File: frontend/src/apis/ragEvaluation.ts
/**
 * RAG Evaluation API client
 * Based on the new RagRecordRef architecture
 */

import { apiUrl } from './base'

// ============ Types ============

export interface RagEvaluationTriggerRequest {
  mode: 'by_kb' | 'by_date_range'
  knowledge_id?: number
  start_date?: string
  end_date?: string
  force?: boolean
}

export interface RagEvaluationTriggerResponse {
  job_id: string
  total_records: number
  pending_evaluation: number
}

export interface RagEvaluationStatusResponse {
  job_id: string
  status: string
  total: number
  completed: number
  failed: number
  skipped: number
}

export interface TrendDataPoint {
  date: string
  total_rag_queries: number
  evaluated_count: number
  evaluation_coverage: number
  pass_count: number
  fail_count: number
  undetermined_count: number
  pass_rate: number | null
  avg_total_score: number | null
  avg_retrieval_score: number | null
  avg_generation_score: number | null
}

export interface TrendComparison {
  pass_rate_change: number | null
  avg_score_change: number | null
}

export interface NonRagStats {
  direct_injection_count: number
  selected_documents_count: number
  non_rag_ratio: number
}

export interface EvaluationTrendsResponse {
  trends: TrendDataPoint[]
  comparison: {
    day_over_day?: TrendComparison
    week_over_week?: TrendComparison
  } | null
  non_rag_stats: NonRagStats | null
}

export interface PeriodStats {
  date_range: string
  evaluated_count: number
  pass_rate: number | null
  avg_total_score: number | null
}

export interface EvaluationCompareResponse {
  period1: PeriodStats
  period2: PeriodStats
  changes: {
    pass_rate_change: number | null
    avg_score_change: number | null
    improvement: boolean
  }
}

export interface LowScoreQueryItem {
  rag_record_ref_id: number
  raw_id: number
  record_date: string | null
  knowledge_id: number | null
  knowledge_name: string | null
  user_prompt: string | null
  evaluation_judgment: string | null
  total_score: number | null
  faithfulness_score: number | null
  trulens_groundedness: number | null
  ragas_query_context_relevance: number | null
  trulens_context_relevance: number | null
  evaluated_at: string | null
}

export interface LowScoreQueriesResponse {
  records: LowScoreQueryItem[]
  total: number
  filters_applied: Record<string, unknown>
}

export interface LowScoreKnowledgeBaseItem {
  knowledge_id: number
  knowledge_name: string | null
  namespace: string | null
  evaluated_count: number
  pass_count: number
  fail_count: number
  undetermined_count: number
  pass_rate: number | null
  fail_rate: number | null
  avg_total_score: number | null
  avg_faithfulness_score: number | null
  avg_trulens_groundedness: number | null
  avg_ragas_query_context_relevance: number | null
  avg_trulens_context_relevance: number | null
}

export interface LowScoreKnowledgeBasesResponse {
  knowledge_bases: LowScoreKnowledgeBaseItem[]
  total: number
}

export interface CoreMetrics {
  faithfulness_score: number | null
  trulens_groundedness: number | null
  ragas_query_context_relevance: number | null
  trulens_context_relevance: number | null
  ragas_context_precision_emb: number | null
}

export interface EvaluationResultData {
  id: number
  evaluation_judgment: string | null
  total_score: number | null
  retrieval_score: number | null
  generation_score: number | null
  core_metrics: CoreMetrics
  threshold: number
  evaluated_at: string | null
}

export interface RawDataPreview {
  user_prompt: string | null
  chunks_count: number | null
  chunks_preview: string[] | null
}

export interface RagRecordEvaluationDetailResponse {
  rag_record_ref_id: number
  raw_id: number
  evaluation_status: string
  evaluation_result: EvaluationResultData | null
  raw_data: RawDataPreview | null
}

export interface KnowledgeBaseEvaluationStats {
  rag_retrieval_count: number
  evaluated_count: number
  evaluation_coverage: number
  pass_count: number
  fail_count: number
  undetermined_count: number
  pass_rate: number | null
  avg_total_score: number | null
}

// ============ API Functions ============

/**
 * Trigger RAG evaluation job
 */
export async function triggerRagEvaluation(
  request: RagEvaluationTriggerRequest
): Promise<RagEvaluationTriggerResponse> {
  const url = apiUrl('/daily/evaluation/trigger')
  const response = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(request),
  })
  if (!response.ok) throw new Error('Failed to trigger evaluation')
  return response.json()
}

/**
 * Get RAG evaluation job status
 */
export async function getRagEvaluationStatus(
  jobId: string
): Promise<RagEvaluationStatusResponse> {
  const url = apiUrl(`/daily/evaluation/status/${jobId}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to get evaluation status')
  return response.json()
}

/**
 * Get evaluation trends
 */
export async function getEvaluationTrends(params?: {
  days?: number
  granularity?: string
  knowledge_id?: number
}): Promise<EvaluationTrendsResponse> {
  const searchParams = new URLSearchParams()
  if (params?.days) searchParams.set('days', params.days.toString())
  if (params?.granularity) searchParams.set('granularity', params.granularity)
  if (params?.knowledge_id) searchParams.set('knowledge_id', params.knowledge_id.toString())

  const url = apiUrl(`/daily/evaluation/trends${searchParams.toString() ? `?${searchParams}` : ''}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to get evaluation trends')
  return response.json()
}

/**
 * Compare evaluation periods
 */
export async function compareEvaluationPeriods(params: {
  period1_start: string
  period1_end: string
  period2_start: string
  period2_end: string
  knowledge_id?: number
}): Promise<EvaluationCompareResponse> {
  const searchParams = new URLSearchParams()
  searchParams.set('period1_start', params.period1_start)
  searchParams.set('period1_end', params.period1_end)
  searchParams.set('period2_start', params.period2_start)
  searchParams.set('period2_end', params.period2_end)
  if (params.knowledge_id) searchParams.set('knowledge_id', params.knowledge_id.toString())

  const url = apiUrl(`/daily/evaluation/compare?${searchParams}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to compare evaluation periods')
  return response.json()
}

/**
 * Get low score queries
 */
export async function getLowScoreQueries(params?: {
  start_date?: string
  end_date?: string
  knowledge_id?: number
  judgment?: string
  sort_by?: string
  limit?: number
  offset?: number
}): Promise<LowScoreQueriesResponse> {
  const searchParams = new URLSearchParams()
  if (params?.start_date) searchParams.set('start_date', params.start_date)
  if (params?.end_date) searchParams.set('end_date', params.end_date)
  if (params?.knowledge_id) searchParams.set('knowledge_id', params.knowledge_id.toString())
  if (params?.judgment) searchParams.set('judgment', params.judgment)
  if (params?.sort_by) searchParams.set('sort_by', params.sort_by)
  if (params?.limit) searchParams.set('limit', params.limit.toString())
  if (params?.offset) searchParams.set('offset', params.offset.toString())

  const url = apiUrl(`/daily/evaluation/low-score/queries${searchParams.toString() ? `?${searchParams}` : ''}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to get low score queries')
  return response.json()
}

/**
 * Get low score knowledge bases
 */
export async function getLowScoreKnowledgeBases(params?: {
  start_date?: string
  end_date?: string
  judgment?: string
  sort_by?: string
  limit?: number
  offset?: number
}): Promise<LowScoreKnowledgeBasesResponse> {
  const searchParams = new URLSearchParams()
  if (params?.start_date) searchParams.set('start_date', params.start_date)
  if (params?.end_date) searchParams.set('end_date', params.end_date)
  if (params?.judgment) searchParams.set('judgment', params.judgment)
  if (params?.sort_by) searchParams.set('sort_by', params.sort_by)
  if (params?.limit) searchParams.set('limit', params.limit.toString())
  if (params?.offset) searchParams.set('offset', params.offset.toString())

  const url = apiUrl(`/daily/evaluation/low-score/knowledge-bases${searchParams.toString() ? `?${searchParams}` : ''}`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to get low score knowledge bases')
  return response.json()
}

/**
 * Get RAG record evaluation detail
 */
export async function getRagRecordEvaluationDetail(
  ragRecordRefId: number
): Promise<RagRecordEvaluationDetailResponse> {
  const url = apiUrl(`/daily/evaluation/rag-records/${ragRecordRefId}/evaluation`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to get RAG record evaluation detail')
  return response.json()
}

/**
 * Get knowledge base evaluation stats
 */
export async function getKnowledgeBaseEvaluationStats(
  kbId: number
): Promise<KnowledgeBaseEvaluationStats> {
  const url = apiUrl(`/daily/evaluation/knowledge-bases/${kbId}/evaluation-stats`)
  const response = await fetch(url)
  if (!response.ok) throw new Error('Failed to get knowledge base evaluation stats')
  return response.json()
}

File: backend/app/schemas/rag_evaluation.py
"""
Schemas for RAG evaluation API endpoints (based on RagRecordRef).
"""
from datetime import date, datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field


# ============ Trigger Evaluation ============

class RagEvaluationTriggerRequest(BaseModel):
    """Request body for triggering RAG evaluation."""

    mode: Literal["by_kb", "by_date_range"]
    knowledge_id: Optional[int] = None  # Required for mode=by_kb
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    force: bool = False  # Force re-evaluation of completed records


class RagEvaluationTriggerResponse(BaseModel):
    """Response for RAG evaluation trigger endpoint."""

    job_id: str
    total_records: int  # Total rag_retrieval records matching criteria
    pending_evaluation: int  # Number of records to be evaluated


class RagEvaluationStatusResponse(BaseModel):
    """Response for RAG evaluation job status."""

    job_id: str
    status: str  # started, running, completed, failed
    total: int
    completed: int
    failed: int
    skipped: int


# ============ Evaluation Trends ============

class TrendDataPoint(BaseModel):
    """Single data point in trend."""

    date: str
    total_rag_queries: int
    evaluated_count: int
    evaluation_coverage: float
    pass_count: int
    fail_count: int
    undetermined_count: int
    pass_rate: Optional[float] = None
    avg_total_score: Optional[float] = None
    avg_retrieval_score: Optional[float] = None
    avg_generation_score: Optional[float] = None


class TrendComparison(BaseModel):
    """Trend comparison data."""

    pass_rate_change: Optional[float] = None
    avg_score_change: Optional[float] = None


class NonRagStats(BaseModel):
    """Non-RAG usage statistics."""

    direct_injection_count: int
    selected_documents_count: int
    non_rag_ratio: float


class EvaluationTrendsResponse(BaseModel):
    """Response for evaluation trends."""

    trends: List[TrendDataPoint]
    comparison: Optional[Dict[str, TrendComparison]] = None
    non_rag_stats: Optional[NonRagStats] = None


# ============ Period Comparison ============

class PeriodStats(BaseModel):
    """Statistics for a time period."""

    date_range: str
    evaluated_count: int
    pass_rate: Optional[float] = None
    avg_total_score: Optional[float] = None


class PeriodChanges(BaseModel):
    """Changes between two periods."""

    pass_rate_change: Optional[float] = None
    avg_score_change: Optional[float] = None
    improvement: bool = False


class EvaluationCompareResponse(BaseModel):
    """Response for period comparison."""

    period1: PeriodStats
    period2: PeriodStats
    changes: PeriodChanges


# ============ Low Score Rankings ============

class LowScoreQueryItem(BaseModel):
    """Single low-score query item."""

    rag_record_ref_id: int
    raw_id: int
    record_date: Optional[str] = None
    knowledge_id: Optional[int] = None
    knowledge_name: Optional[str] = None
    user_prompt: Optional[str] = None
    evaluation_judgment: Optional[str] = None
    total_score: Optional[float] = None
    faithfulness_score: Optional[float] = None
    trulens_groundedness: Optional[float] = None
    ragas_query_context_relevance: Optional[float] = None
    trulens_context_relevance: Optional[float] = None
    evaluated_at: Optional[datetime] = None


class LowScoreQueriesResponse(BaseModel):
    """Response for low-score queries."""

    records: List[LowScoreQueryItem]
    total: int
    filters_applied: Dict[str, Any]


class LowScoreKnowledgeBaseItem(BaseModel):
    """Single low-score knowledge base item."""

    knowledge_id: int
    knowledge_name: Optional[str] = None
    namespace: Optional[str] = None
    evaluated_count: int
    pass_count: int
    fail_count: int
    undetermined_count: int
    pass_rate: Optional[float] = None
    fail_rate: Optional[float] = None
    avg_total_score: Optional[float] = None
    avg_faithfulness_score: Optional[float] = None
    avg_trulens_groundedness: Optional[float] = None
    avg_ragas_query_context_relevance: Optional[float] = None
    avg_trulens_context_relevance: Optional[float] = None


class LowScoreKnowledgeBasesResponse(BaseModel):
    """Response for low-score knowledge bases."""

    knowledge_bases: List[LowScoreKnowledgeBaseItem]
    total: int


# ============ Single Record Evaluation Detail ============

class CoreMetrics(BaseModel):
    """Core evaluation metrics."""

    faithfulness_score: Optional[float] = None
    trulens_groundedness: Optional[float] = None
    ragas_query_context_relevance: Optional[float] = None
    trulens_context_relevance: Optional[float] = None
    ragas_context_precision_emb: Optional[float] = None


class EvaluationResultData(BaseModel):
    """Evaluation result data."""

    id: int
    evaluation_judgment: Optional[str] = None
    total_score: Optional[float] = None
    retrieval_score: Optional[float] = None
    generation_score: Optional[float] = None
    core_metrics: CoreMetrics
    threshold: float = 0.6
    evaluated_at: Optional[datetime] = None


class RawDataPreview(BaseModel):
    """Raw data preview."""

    user_prompt: Optional[str] = None
    chunks_count: Optional[int] = None
    chunks_preview: Optional[List[str]] = None


class RagRecordEvaluationDetailResponse(BaseModel):
    """Response for single RAG record evaluation detail."""

    rag_record_ref_id: int
    raw_id: int
    evaluation_status: str
    evaluation_result: Optional[EvaluationResultData] = None
    raw_data: Optional[RawDataPreview] = None


# ============ Query List Enhancement ============

class RagQueryItem(BaseModel):
    """Enhanced query item with evaluation info."""

    id: int
    raw_id: int
    record_date: Optional[str] = None
    context_type: Optional[str] = None
    injection_mode: Optional[str] = None
    knowledge_id: Optional[int] = None
    knowledge_name: Optional[str] = None
    namespace: Optional[str] = None
    query: Optional[str] = None
    chunks_count: Optional[int] = None
    created_at: Optional[str] = None

    # Evaluation fields
    is_evaluable: bool = False
    evaluation_status: Optional[str] = None
    evaluation_judgment: Optional[str] = None
    total_score: Optional[float] = None
    evaluated_at: Optional[datetime] = None


class RagQueryListResponse(BaseModel):
    """Response for RAG query list."""

    items: List[RagQueryItem]
    total: int
    page: int
    page_size: int


# ============ Knowledge Base Evaluation Stats ============

class KnowledgeBaseEvaluationStats(BaseModel):
    """Evaluation statistics for a knowledge base."""

    rag_retrieval_count: int
    evaluated_count: int
    evaluation_coverage: float
    pass_count: int
    fail_count: int
    undetermined_count: int
    pass_rate: Optional[float] = None
    avg_total_score: Optional[float] = None
